interceptors/logging/interceptors.go

```diff
@@ -22,65 +22,132 @@ func extractFields(tags tags.Tags) Fields {
 }
 
 type reporter struct {
-	ctx             context.Context
-	typ             interceptors.GRPCType
-	service, method string
-
-	opts   *options
-	logger Logger
+	ctx                        context.Context
+	typ                        interceptors.GRPCType
+	service, method            string
+	firstRequestMessageLogged  bool
+	firstResponseMessageLogged bool
+	isServer                   bool
+	opts                       *options
+	logger                     Logger
 
 	kind string
 }
 
+func (c *reporter) addMetaDetails(logger Logger) Logger {
+	if c.isServer {
+		logger = logger.With("component", "server")
+	}
+	if !c.isServer {
+		logger = logger.With("component", "client")
+	}
+	// Attach the kind, protocol and type as a structured log entry.
+	logger = logger.With("kind", c.kind)
+	logger = logger.With("type", string(c.typ))
+	logger = logger.With("protocol", "grpc")
+
+	return logger
+}
+
+// PostCall logs the server/method name details and error if any.
 func (c *reporter) PostCall(err error, duration time.Duration) {
 	if !c.opts.shouldLog(interceptors.FullMethod(c.service, c.method), err) {
 		return
 	}
 	if err == io.EOF {
 		err = nil
 	}
+	logger := c.addMetaDetails(c.logger)
 	// Get optional, fresh tags.
-	logger := c.logger.With(extractFields(tags.Extract(c.ctx))...)
-
+	logger = logger.With(extractFields(tags.Extract(c.ctx))...)
 	code := c.opts.codeFunc(err)
 	logger = logger.With("grpc.code", code.String())
 	if err != nil {
-		logger = logger.With("error", fmt.Sprintf("%v", err))
+		logger = logger.With("grpc.error", fmt.Sprintf("%v", err))
+	}
+	logger = logger.With(c.opts.durationFieldFunc(duration)...)
+	logger.Log(c.opts.levelFunc(code), "finished call")
+}
+
+// PostMsgSend logs the details of the servingObject that is flowing out of the rpc.
+// resp object wrt server.
+// Log the details of the first request, skip if the object is response.
+func (c *reporter) PostMsgSend(resp interface{}, err error, duration time.Duration) {
+	if !c.opts.shouldLog(interceptors.FullMethod(c.service, c.method), err) {
+		return
+	}
+
+	// If the first message is logged skip the rest of the logging.
+	// If the serving object is response, skip the logging.
+	if c.firstResponseMessageLogged || !c.isServer {
```

Comment:
What about one bool flag? `startedCallLogged bool` ? (: for server and client the first PostMsgSend and PostMsgRecv being called "wins"

Reply:
Changed